### PR TITLE
Attend and excite 2

### PR DESCRIPTION
**Pull request by yiyixuxu**
open a new PR since I messed up the original PR 
This PR originated from https://github.com/huggingface/diffusers/pull/2275 with full credit going to @evinpinar as the original author of the pipeline 

How to use 
```python
import torch
from diffusers import StableDiffusionAttendAndExcitePipeline

pipe = StableDiffusionAttendAndExcitePipeline.from_pretrained(
     "CompVis/stable-diffusion-v1-4", torch_dtype=torch.float16
).to("cuda")

pipe = pipe.to("cuda")

prompt = "a painting of an elephant with glasses"
indices_to_alter = [5, 7]
seed = 51
g = torch.Generator("cuda").manual_seed(seed)

images = pipe(
    prompt= prompt,
    indices_to_alter=indices_to_alter,
    guidance_scale=7.5,
    generator=g,
    num_inference_steps=50,
    max_iter_to_alter=25,
    ).images

image = images[0]
image.save(f"../images/test2_{prompt}_{seed}.png")
```

test2_a painting of an elephant with glasses_51

### CODE REVIEW

File: docs/source/en/api/pipelines/stable_diffusion/attend_and_excite.mdx [hunk 1]
```diff
@@ -0,0 +1,71 @@
+<!--Copyright 2022 The HuggingFace Team. All rights reserved.
```
Comment (patil-suraj) [inline docs/source/en/api/pipelines/stable_diffusion/attend_and_excite.mdx:1]:
```suggestion
<!--Copyright 2024 The HuggingFace Team. All rights reserved.
```

File: docs/source/en/api/pipelines/stable_diffusion/attend_and_excite.mdx [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+pipe = StableDiffusionAttendAndExcitePipeline.from_pretrained(model_id, torch_dtype=torch.float16).to("cuda")
+pipe = pipe.to("cuda")
+
+prompt = "a cat and a frog"
+indices = [2, 5]
+seed = 6141
+generator = torch.Generator("cuda").manual_seed(seed)
+
+images = pipe(
+    prompt=[prompt],
```
Comment (patil-suraj) [inline docs/source/en/api/pipelines/stable_diffusion/attend_and_excite.mdx:55]:
```suggestion
    prompt=prompt,
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_attend_and_excite.py [hunk 3]
```diff
@@ -0,0 +1,993 @@
+# Copyright 2022 The HuggingFace Team. All rights reserved.
```
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_attend_and_excite.py:1]:
```suggestion
# Copyright 2023 The HuggingFace Team. All rights reserved.
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_attend_and_excite.py [hunk 4]
```diff
@@ -0,0 +1,993 @@
...
+    Examples:
+        ```py
+        >>> import torch
+        >>> from diffusers import StableDiffusionAttendAndExcitePipeline
+
+        >>> pipe = StableDiffusionAttendAndExcitePipeline.from_pretrained(
+        ...     "CompVis/stable-diffusion-v1-4", torch_dtype=torch.float16
+        ... ).to("cuda")
+
+        >>> pipe = pipe.to("cuda")
```
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_attend_and_excite.py:45]:
```suggestion
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_attend_and_excite.py [hunk 5]
```diff
@@ -0,0 +1,993 @@
...
+
+        >>> prompt = "a cat and a frog"
+        >>> indices = [2, 5]
+        >>> seed = 6141
+        >>> g = torch.Generator("cuda").manual_seed(seed)
+
+        >>> images = pipe(
+        ...     prompt=[prompt],
```
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_attend_and_excite.py:53]:
```suggestion
        ...     prompt=prompt,
```

File: tests/pipelines/stable_diffusion_2/test_stable_diffusion_attend_and_excite.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
+# coding=utf-8
+# Copyright 2022 HuggingFace Inc.
```
Comment (patil-suraj) [inline tests/pipelines/stable_diffusion_2/test_stable_diffusion_attend_and_excite.py:2]:
```suggestion
# Copyright 2023 HuggingFace Inc.
```

File: docs/source/en/api/pipelines/stable_diffusion/attend_and_excite.mdx [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+```python
+import torch
+from diffusers import StableDiffusionAttendAndExcitePipeline
+
+model_id = "CompVis/stable-diffusion-v1-4"
+pipe = StableDiffusionAttendAndExcitePipeline.from_pretrained(model_id, torch_dtype=torch.float16).to("cuda")
+pipe = pipe.to("cuda")
+
+prompt = "a cat and a frog"
+indices = [2, 5]
```
Comment (patil-suraj) [inline docs/source/en/api/pipelines/stable_diffusion/attend_and_excite.mdx:50]: Would be cool to document what are `indices` and how to get these indices.

**Review by patil-suraj — APPROVED** [review]
Looks great to me, thanks a lot for adding this new pipeline! The API is very clean!

Mostly left some nits, my main comment is, would be nice to document how to get the indices.

File: docs/source/en/api/pipelines/stable_diffusion/attend_and_excite.mdx (hunk 1 above)
Comment (patrickvonplaten) [reply docs/source/en/api/pipelines/stable_diffusion/attend_and_excite.mdx:1]: living in the future

Comment (patrickvonplaten) [reply docs/source/en/api/pipelines/stable_diffusion/attend_and_excite.mdx:1]: Should probs be 2023

File: docs/source/en/api/pipelines/stable_diffusion/attend_and_excite.mdx (hunk 7 above)
Comment (patrickvonplaten) [reply docs/source/en/api/pipelines/stable_diffusion/attend_and_excite.mdx:50]: Also those correspond to words no? Maybe call it `word_indices`?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_attend_and_excite.py [hunk 8]
```diff
@@ -0,0 +1,993 @@
...
+        >>> pipe = StableDiffusionAttendAndExcitePipeline.from_pretrained(
+        ...     "CompVis/stable-diffusion-v1-4", torch_dtype=torch.float16
+        ... ).to("cuda")
+
+        >>> pipe = pipe.to("cuda")
+
+        >>> prompt = "a cat and a frog"
+        >>> indices = [2, 5]
+        >>> seed = 6141
+        >>> g = torch.Generator("cuda").manual_seed(seed)
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_attend_and_excite.py:50]:
```suggestion
        >>> generator = torch.Generator("cuda").manual_seed(seed)
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_attend_and_excite.py [hunk 9]
```diff
@@ -0,0 +1,993 @@
...
+
+        self.unet.set_attn_processor(attn_procs)
+        self.attention_store.num_att_layers = cross_att_count
+
+    @torch.no_grad()
+    @replace_example_docstring(EXAMPLE_DOC_STRING)
+    def __call__(
+        self,
+        prompt: Union[str, List[str]],
+        indices: List[int],
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_attend_and_excite.py:662]:
```suggestion
        token_indices: List[int],
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_attend_and_excite.py [hunk 10]
```diff
@@ -0,0 +1,993 @@
...
+        scale_factor: int = 20,
+    ):
+        r"""
+        Function invoked when calling the pipeline for generation.
+
+        Args:
+            prompt (`str` or `List[str]`, *optional*):
+                The prompt or prompts to guide the image generation. If not defined, one has to pass `prompt_embeds`.
+                instead.
+            indices (`List[int]`):
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_attend_and_excite.py:690]:
```suggestion
            token_indices (`List[int]`):
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_attend_and_excite.py [hunk 11]
```diff
@@ -0,0 +1,993 @@
...
+            else:
+                continue
+
+            cross_att_count += 1
+            attn_procs[name] = AttendExciteCrossAttnProcessor(
+                attnstore=self.attention_store, place_in_unet=place_in_unet
+            )
+
+        self.unet.set_attn_processor(attn_procs)
+        self.attention_store.num_att_layers = cross_att_count
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_attend_and_excite.py:655]: Could we add a new method, called:

`get_token_indices`

that returns the input prompt and the corresponding token indices, this should help the user to know which tokens to use, e.g.:

```suggestion
        self.attention_store.num_att_layers = cross_att_count
        
    def get_indices(prompt: str) -> Dict[str, int]:
        ids = self.tokenizer(prompt).input_ids
        
        indices = {tok: i for tok, i in zip(self.tokenizer.convert_ids_to_tokens(ids), ids)}
        return indices
```

Then people can call this before to know which indices to pass.

**Comment by patrickvonplaten** [comment]
Cool, the PR looks nice, I'd maybe add one utility function `get_indices` that the user can call with the prompt to know exactly which indices to pass to the pipeline.

File: docs/source/en/api/pipelines/stable_diffusion/attend_and_excite.mdx (hunk 7 above)
Comment (yiyixuxu) [reply docs/source/en/api/pipelines/stable_diffusion/attend_and_excite.mdx]: these are just positions of the tokens in the sentence - I added the function you suggested and changed the name

File: docs/source/en/api/pipelines/stable_diffusion/attend_and_excite.mdx (hunk 1 above)
Comment (yiyixuxu) [reply docs/source/en/api/pipelines/stable_diffusion/attend_and_excite.mdx]: 2023 it is 😂

File: docs/source/en/api/pipelines/stable_diffusion/attend_and_excite.mdx [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+from diffusers import StableDiffusionAttendAndExcitePipeline
+
+model_id = "CompVis/stable-diffusion-v1-4"
+pipe = StableDiffusionAttendAndExcitePipeline.from_pretrained(model_id, torch_dtype=torch.float16).to("cuda")
+pipe = pipe.to("cuda")
+
+prompt = "a cat and a frog"
+
+# use get_indices function to find out indices of the tokens you want to alter
+pipe.get_indices(prompt)
```
Comment (patrickvonplaten) [inline docs/source/en/api/pipelines/stable_diffusion/attend_and_excite.mdx:52]: an we display the dict here as a comment?

**Review by patrickvonplaten — APPROVED** [review]
Awesome! Feel free to merge whenever :-)

**Comment by Skquark** [comment]
Trying to get my implementation working, but I keep getting this error:
```python
Traceback (most recent call last):
  File "<ipython-input-4-e93657c66c31>", line 10052, in start_diffusion
    images = pipe_attend_and_excite(prompt=ptext, token_indices=token_indices, max_iter_to_alter=int(prefs['max_iter_to_alter']), height=arg['height'], width=arg['width'], num_inference_steps=arg['steps'], guidance_scale=arg['guidance_scale'], eta=arg['eta'], generator=generator, callback=callback_fn, callback_steps=1).images
  File "/usr/local/lib/python3.8/dist-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_attend_and_excite.py", line 875, in __call__
    max_attention_per_index = self._aggregate_and_get_max_attention_per_token(
  File "/usr/local/lib/python3.8/dist-packages/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_attend_and_excite.py", line 564, in _aggregate_and_get_max_attention_per_token
    attention_maps = self.attention_store.aggregate_attention(
  File "/usr/local/lib/python3.8/dist-packages/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_attend_and_excite.py", line 101, in aggregate_attention
    out = torch.cat(out, dim=0)
RuntimeError: torch.cat(): expected a non-empty list of Tensors
```
I've tried changing the way I'm calling the pipeline, double checked my indices, constructor, and everything else and can't figure out how to get past it.  My code is pretty much the same as the example.  Is this a bug or could I be doing something wrong?

**Comment by Skquark** [comment]
I stripped back my code piece by piece to debug error, and finally I discovered it was because I was adding width and height to the pipe, even though it was just width=640, height=640 I was testing.  So does this pipeline only make 512x512 images?  Can we expect that limitation to change, or just live with that size? If so, maybe you can put in something to ignore those parameters.
Now got it included in my StableDiffusionDeluxe.com app, set it up to include a plus before the word to mark as a token, so prompt="a +cat and a +frog" makes token_indices=[2, 5] automatically.  Works well, been getting surprisingly awesome images out of it already.. Thanks...

**Comment by polavishnu4444** [comment]
Is this going to work for 2.1 as well?

**Pull request merged**